### PR TITLE
[AIRFLOW-1325] Add ElasticSearch log handler and reader

### PR DESCRIPTION
**Pull request by KevinYang21**
### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-1325) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1325

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
- Add ElasticSearch log handler which writes to local file and reads from ElasticSearch (need seperate EFK/ELK stack to have the local file uploaded to ElasticSearch).
- Add support of auto-tailing task instance log, when user scroll to the bottom.
- Log reading major changes:
    - No log will be loaded before rendering ti_log.html, instead the html will make calls, potentially repeatedly, to new get_logs_with_metadata() endpoint to load logs (see the graphs below).
    - FileTaskHandler().read() would now take metadata as an option parameter to enable streaming read and return (logs, metadatas) tuple, which may contain end of log and next offset info.

Note that codes under /tests/utils/log/elasticmock are mostly copy/paste from ElasticMock with some compatibility changes and License headers added.

A simple high level overview of the architecture change of the task instance log loading:
old_arch
new_arch

In the new architecture, metadata would be used to mark end of log, current offset in stream, etc.

Here are some screenshot for the new loading icon, which will be displayed as long as the task log page is still loading log (e.g. loading large log or loading streaming log):
screen shot 2018-04-12 at 12 52 07 am
screen shot 2018-04-12 at 12 59 53 am

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests/utils/log/test_es_task_handler.py
tests/www/test_views.py: test_get_logs_with_metadata

Similar logic has been running in our branch for more than half a year. Tested Elasticsearch as logging backend on our cluster end to end.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -857,7 +857,7 @@ def __init__(self, task, execution_date, state=None):
         self.init_on_load()
         # Is this TaskInstance being currently running within `airflow run --raw`.
         # Not persisted to the database so only valid for the current process
-        self.is_raw = False
+        self.raw = False
```
Comment (KevinYang21) [inline airflow/models.py:860]: only self.raw is used and set outside constructor.

File: airflow/models.py [hunk 2]
```diff
@@ -1956,8 +1956,8 @@ def init_run_context(self, raw=False):
         """
         Sets the log context.
         """
-        self._set_context(self)
         self.raw = raw
+        self._set_context(self)
```
Comment (KevinYang21) [inline airflow/models.py:1960]: raw is used by custom log handler when setting context, e.g. S3TaskHandler use it to determine if it should upload log on close.

File: airflow/www/views.py [hunk 3]
```diff
@@ -719,6 +718,66 @@ def rendered(self):
             form=form,
             title=title, )
 
+    @expose('/get_logs_with_metadata')
+    @login_required
+    @wwwutils.action_logging
+    @provide_session
+    def get_logs_with_metadata(self, session=None):
```
Comment (KevinYang21) [inline airflow/www/views.py:725]: Not putting the endpoint into API to reuse the dagbag. Open for discussion on this.

**Comment by KevinYang21** [comment]
@aoen @saguziel @Fokko PTAL

File: airflow/utils/log/wasb_task_handler.py [hunk 4]
```diff
@@ -1,175 +1,177 @@
...
-        try:
-            self.hook.load_string(
-                log,
-                self.wasb_container,
-                remote_log_location,
-            )
-        except AzureHttpError:
-            self.log.exception('Could not write logs to %s',
-                               remote_log_location)
+# -*- coding: utf-8 -*-
```
Comment (Fokko) [inline airflow/utils/log/wasb_task_handler.py:1]: This is line-endings I assume.

File: airflow/utils/log/es_task_handler.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        :param log_id_template: log id template
+        :param host: Elasticsearch host name
+        """
+        super(ElasticsearchTaskHandler, self).__init__(
+            base_log_folder, filename_template)
+        self.log_id_template = log_id_template
+        self.log_id_jinja_template = None
+        self.closed = False
+
+        if "{{" in self.log_id_template:  # jinja mode
```
Comment (Fokko) [inline airflow/utils/log/es_task_handler.py]: We might want to move this template logic in one single place, it is also in multiple task_handlers.

**Review by Fokko — APPROVED** [review]

File: airflow/utils/log/es_task_handler.py (hunk 5 above)
Comment (KevinYang21) [reply airflow/utils/log/es_task_handler.py]: I think the only other two places using similar logic is in FileTaskHandler and FileProcessorHandler, which constructs the filename template. I think I can extract the template construction logic and make both log_id template and filename template construction use it, sounds good?

File: airflow/utils/log/wasb_task_handler.py (hunk 4 above)
Comment (KevinYang21) [reply airflow/utils/log/wasb_task_handler.py:1]: Your are perfectly correct, the file was using DOS style line-endings and now unifying it to use UNIX style, hope this is fine.

**Review by Fokko — APPROVED** [review]
LGTM, thanks

**Comment by ashb** [comment]
@KevinYang21 

> Note that codes under /tests/utils/log/elasticmock are mostly copy/paste from ElasticMock with some compatibility changes and License headers added.

What were the compatability changes needed? It would be good if we could remove this vendored dep

**Comment by KevinYang21** [comment]
@ashb sorry somehow I lost track to this conversation. I should have provided more details there my bad. So the compatibility changes needed are basically [this method](https://github.com/apache/incubator-airflow/blob/master/tests/utils/log/elasticmock/fake_elasticsearch.py#L283-L295). It looks like the doc_type type issue has been fixed tho in [this commit](https://github.com/vrcmarcos/elasticmock/commit/8895aed57562f6ad7df817c352035eafc12aad60). I can experiment if we can just use the lib w/o modifications( maybe later this week or weekend).

Tho this would mean we still depend on this package, is this what you wanted? Or you actually meant something else when saying `remove this vendored dep`?

**Comment by ashb** [comment]
By "remove this vendored dep" I mean can we remove tests/utils/log/elasticmock.py and pull in the version from pypi (I was looking at our licenses and removing this one would just make it a small bit simpler is all)

**Comment by KevinYang21** [comment]
@ashb Got it, I'll be working on it this week.

**Pull request closed without merging**